tiledb/sm/query/query.cc

```diff
@@ -2264,78 +2272,41 @@ void Query::set_status(QueryStatus status) {
+      break;
 
-  if (!array_schema_->domain().all_dims_fixed())
-    return logger_->status(
-        Status_QueryError("Cannot set subarray; Function not applicable to "
-                          "domains with variable-sized dimensions"));
+    case QueryType::WRITE:
+    case QueryType::MODIFY_EXCLUSIVE:
+      if (!array_schema_->dense()) {
+        throw QueryStatusException(
+            "Cannot set subarray; Setting a subarray is not supported on "
+            "sparse writes.");
+      }
+      break;
 
-  // Prepare a subarray object
-  Subarray sub(array_, layout_, stats_, logger_);
-  if (subarray != nullptr) {
-    auto dim_num = array_schema_->dim_num();
-    auto s_ptr = (const unsigned char*)subarray;
-    uint64_t offset = 0;
-
-    bool err_on_range_oob = true;
-    if (type_ == QueryType::READ) {
-      // Get read_range_oob config setting
-      bool found = false;
-      std::string read_range_oob_str =
-          config().get("sm.read_range_oob", &found);
-      assert(found);
-      if (read_range_oob_str != "error" && read_range_oob_str != "warn")
-        return logger_->status(Status_QueryError(
-            "Invalid value " + read_range_oob_str +
-            " for sm.read_range_obb. Acceptable values are 'error' or "
-            "'warn'."));
-      err_on_range_oob = read_range_oob_str == "error";
-    }
+    default:
 
-    for (unsigned d = 0; d < dim_num; ++d) {
-      auto r_size{2 * array_schema_->dimension_ptr(d)->coord_size()};
-      Range range(&s_ptr[offset], r_size);
-      RETURN_NOT_OK(sub.add_range(d, std::move(range), err_on_range_oob));
-      offset += r_size;
-    }
+      throw QueryStatusException(
+          "Cannot set subarray; Setting a subarray is not supported for query "
+          "type '" +
+          query_type_str(type_) + "'.");
   }
 
-  if (type_ == QueryType::WRITE || type_ == QueryType::MODIFY_EXCLUSIVE) {
-    // Not applicable to sparse arrays
-    if (!array_schema_->dense())
-      return logger_->status(Status_WriterError(
-          "Setting a subarray is not supported in sparse writes"));
-
-    // Subarray must be unary for dense writes
-    if (sub.range_num() != 1)
-      return logger_->status(
-          Status_WriterError("Cannot set subarray; Multi-range dense writes "
-                             "are not supported"));
-
-    // When executed from serialization, resetting the strategy will delete
-    // the fragment for global order writes. We need to prevent this, thus
-    // persist the fragment between remote global order write submits.
-    if (strategy_ != nullptr && layout_ != Layout::GLOBAL_ORDER) {
-      strategy_->reset();
-    }
+  // Check this isn't an already initialized query using dimension labels.
+  if (status_ != QueryStatus::UNINITIALIZED && uses_dimension_labels()) {
+    throw QueryStatusException(
+        "Cannot set subarray; Setting a subarray on a query that uses "
+        "dimension labels is not supported.");
   }
```

Comment:
Possibly confusing message content (or am I not following logic)?  
Appears that if status == UNINITIALIZED then it will allow setting the query, 
so user's seeing message might wonder why not, they've been doing set_subarray() without issues...
'...; Replacing a subarray...', '...; Overwriting a subarray...',
'...; Setting a subarray on an active[what descriptive state appropriate?] query...
or ...'?
(similar with other set_subarray)

Reply:
Thinking a bit more - is it also illegal to set it on a COMPLETEd query?
Does prohibiting setting on COMPLETEd query inhibit re-using query (objects), are there use cases where objects might be re-used without otherwise being freshly allocated, is there a reset()?
Is this a new prohibition, guess you do have the PR marked as BREAKING_BEHAVIOUR ...?

Reply:
You right, the message is wrong. With regards to you other question, it should fail for completed queries too, which does in fact prevent re-using the query object.